### PR TITLE
Ins 192 smart contract experiment

### PR DESCRIPTION
**Pull request by ilyapt**
**- What I did**

- New smart contracts without interfaces
- AddChild (foundation)
- FakeContext injection (foundation)

**- How to verify it**

- run tests

### CODE REVIEW

File: genesis/experiment/wallet/wallet.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package wallet
+
+import (
+	"github.com/insolar/insolar/genesis/experiment/allowance"
+	"github.com/insolar/insolar/toolkit/go/foundation"
+)
+
+// todo make this investigation through reflection
```
Comment (golangci) [inline genesis/experiment/wallet/wallet.go]: comment on exported var TypeReference should be of the form "TypeReference ..."

File: genesis/experiment/wallet/wallet.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+func (w *Wallet) Receive(amount uint, from *foundation.Reference) {
+	intr := foundation.GetObject(from)
+	fromWallet := intr.GetImplementationFor(&TypeReference).(*Wallet)
+
+	Allowance := fromWallet.Allocate(amount, w.GetContext().Me)
+	w.balance += Allowance.TakeAmount()
+}
+
+func (w *Wallet) GetTotalBalance() uint {
+	var totalAllowanced uint = 0
```
Comment (golangci) [inline genesis/experiment/wallet/wallet.go]: should drop = 0 from declaration of var totalAllowanced; it is the zero value

File: toolkit/go/foundation/foundation.go [hunk 3]
```diff
@@ -36,30 +49,87 @@ func (bc *BaseContract) SetContext(c *CallContext) {
...
 
-var FakeLedger = make(map[*Reference]interface{})
-var FakeDelegates = make(map[*Reference]map[*Reference]interface{})
-var FakeChildren = make(map[*Reference]map[*Reference][]interface{})
+var FakeLedger = make(map[*Reference]BaseContractInterface)
+var FakeDelegates = make(map[*Reference]map[*Reference]BaseContractInterface)
+var FakeChildren = make(map[*Reference]map[*Reference][]BaseContractInterface)
+
+var FakeContexts = make(map[uint]*CallContext)
+var contextStep uint = 0
```
Comment (golangci) [inline toolkit/go/foundation/foundation.go]: should drop = 0 from declaration of var contextStep; it is the zero value

File: genesis/experiment/wallet/wallet.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	toWallet := foundation.GetImplementationFor(to, &TypeReference).(*Wallet)
+	toWallet.Accept(&a)
+}
+
+func (w *Wallet) Accept(a *allowance.Allowance) {
+	w.balance += a.TakeAmount()
+}
+
+func (w *Wallet) GetTotalBalance() uint {
+	var totalAllowanced uint = 0
```
Comment (golangci) [inline genesis/experiment/wallet/wallet.go]: should drop = 0 from declaration of var totalAllowanced; it is the zero value

File: toolkit/go/foundation/foundation.go [hunk 5]
```diff
@@ -36,45 +83,106 @@ func (bc *BaseContract) SetContext(c *CallContext) {
...
 
-var FakeLedger = make(map[*Reference]interface{})
-var FakeDelegates = make(map[*Reference]map[*Reference]interface{})
-var FakeChildren = make(map[*Reference]map[*Reference][]interface{})
+var FakeLedger = make(map[string]BaseContractInterface)
+var FakeDelegates = make(map[string]map[string]BaseContractInterface)
+var FakeChildren = make(map[string]map[string][]BaseContractInterface)
+
+var FakeContexts = make(map[uint]*CallContext)
+var contextStep uint = 0
```
Comment (golangci) [inline toolkit/go/foundation/foundation.go]: should drop = 0 from declaration of var contextStep; it is the zero value

File: genesis/experiment/wallet/wallet.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"github.com/insolar/insolar/genesis/experiment/allowance"
+	"github.com/insolar/insolar/toolkit/go/foundation"
+)
+
+// todo make this investigation through reflection
+var TypeReference = foundation.Reference("wallet")
+
+type Wallet struct {
```
Comment (LinnikD) [inline genesis/experiment/wallet/wallet.go]: Should we use any delegate marker?

File: genesis/experiment/member/member.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func (m *Member) GetName() string {
+	return m.Name
+}
+func (m *Member) GetPublicKey() []byte {
+	return m.PublicKey
+}
+
+func NewMember(name string) (*Member, foundation.Reference) {
```
Comment (Vany) [inline genesis/experiment/member/member.go]: Please, just make an object in constructor, we will store it in StoreToLedger() separately

**Pull request merged**